Allows for compatibility with BLF

TACambron opened the pull request:
Added updates throughout the code to update each phone's status in real time so that BLF can be properly used on the phones.

Commit:
Allows for dialing of multiple parties.

Added code primarily to the originateChannel method to allow for
multiple hardcoded phones to be dialed upon StasisStart. When one
dialed phone answers, the other dialed phones are hung up. If all
dialed phones hangup or do not answer, the inbound channel is hungup.
Added two more tests for functionality.

Commit:
Added BLF feature to SLA configuration

Added the ability for the sla application to control the device
state of the devices in the shared extension through ARI. Device state
options include "RINGING", "INUSE", and "NOTINUSE". Testing will also
cover BLF implementation.

Commit:
Progress placeholder. Tests are not complete.

TACambron committed:
Added BLF functionality and mock tests to verify.

Added device state updates so that the stations in an SLA setup
would be able to correctly display their own states for the use of
BLF. Added four tests to cover new features. Also rebased branch off
of the most recent update to original master branch.

TACambron committed:
Temporary fix with last test sometimes failing

Last test will sometimes fail because of timing problems.

TACambron committed:
Broken Placeholder.

TACambron committed:
Added compatibility for a configuration file.

Restructured the original files as well as added new files to
allow for the use of a configuration file so that customization
is more practical.

TACambron committed:
Fixed unintended changes from using the swap function.

The asyncDelay value had been greatly increased on accident but
after reverting it back, tests complete much quicker.

TACambron committed:
Fixed minor syntax issues.

matt-jordan commented on lib/sla.js:
```diff
@@ -116,6 +116,8 @@ function originateChannel(client, inbound, bridge, trunkData) {
   function dialedEnteredStasis(event, line) {
     var answer = Q.denodeify(line.answer.bind(line));
     answer();
+    var deviceState = Q.denodeify(client.deviceStates.update.bind(client));
```
From a readability perspective, put some whitespace between logical groupings of code:

```
var answer = ...
answer()

var deviceState = ...
deviceState(...)

var toKill = ...
```

etc.

matt-jordan replied:
Note that this comment applies throughout this review. You've got a lot of code bunched up, without much in the way marking logical separation between actions being taken.

matt-jordan commented on lib/sla.js:
```diff
@@ -125,6 +127,9 @@ function originateChannel(client, inbound, bridge, trunkData) {
       var hangup = Q.denodeify(unanswered.channel.hangup.bind(
           unanswered.channel));
       hangup();
+      var deviceState = Q.denodeify(client.deviceStates.update.bind(client));
```
This pattern of getting the device state from the client, setting up the device to change, and changing its state, is pretty common throughout this patch.

I'd write a small function that does just that, taking in the name of the device to change and the state to set it to.

matt-jordan commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
       }, asyncDelay);
     } 
   });
+  it('should mark a device as RINGING when dialing an outbound channel',
```
Vertical whitespace is your friend!

matt-jordan commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
+      function(done) {
+    var client = getMockClient();
+    var inbound = getMockChannel();
+    inbound.inbound = true;
```
From a naming perspective, `inbound.inbound` is pretty weird. I'd name the channel something other than `inbound` to avoid the duplication.

samuelg commented on tests/sla_tests.js:
```diff
@@ -82,10 +84,42 @@ var getMockClient = function() {
       return newChan;
     };
+    this.deviceStates = {
+      update: function(param, cb) {
```
'param' should really be 'params' as this is an object. Another common name for such objects in JavaScript is 'options'.

samuelg commented on tests/sla_tests.js:
```diff
@@ -82,10 +84,42 @@ var getMockClient = function() {
     };
+    this.deviceStates = {
+      update: function(param, cb) {
+        var exists = mockDeviceStates.filter(function(deviceState) {
```
'exists' implies this is a boolean. I would rename it to something else since it's used later in to establish whether or not the array is empty.

Another options is to do:

var exists = !mockDeviceStates.filter(function(deviceState) {
  return deviceState.deviceName === param.deviceName;
}).length;

but that is definitely harder to read.

The best option is to use a library like https://lodash.com to overcome some of JavaScript limitations. You could can do something like:

var exists = _.some(mockDeviceStates, function(deviceState) {
  return deviceState.deviceName === param.deviceName;
});

samuelg commented on tests/sla_tests.js:
```diff
@@ -231,7 +267,7 @@ describe('SLA Bridge and Channels Tester', function() {
       .done();
 
-    bridges.push(getMockBridge({type: 'mixing', name: '42'}, function(){}));
+    bridges.push(getMockBridge({type: 'mixing', name: '42'}, function() {}));
```
I'm not sure why there's a function being passed to getMockBridge. That function doesn't expect more than 1 parameter.

samuelg commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
+    var client = getMockClient();
+    var inbound = getMockChannel();
+    inbound.inbound = true;
+    bridges.push(getMockBridge({type: 'mixing', name: '42'}, function() {}));
```
Same here as far as passing a function to getMockBridge.

samuelg commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
+    noEndpoints();
+    function noEndpoints() {
+      setTimeout(function() {
+        if (noEndpoints) {
```
noEndpoints doesn't appear to be defined so it should always be undefined which will be evaluated to false.

samuelg commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
+    invalidConfigurationFile();
+    function invalidConfigurationFile() {
+      setTimeout(function() {
+        if (configurationFailed) {
```
Typically, using setTimeout to check local variables for state during a test is done because you have no direct control over the state you want to check against.

It would be better here to simply provide your own anonymous function to the catch handler instead of relying on errHandler. You could then call done() inside your anonymous function if the error thrown is the one you expect.

samuelg commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
+    var sla = require('../lib/sla.js')(client, config, inbound, '42')
+      .catch(errHandler)
+      .done();
+    client.deviceStates.update({deviceName: 'Stasis:42',
```
This is hard to read, you should go for one of the approached below:

client.deviceStates.update({
  deviceName: 'Stasis:42',
  deviceState: 'NOT_INUSE'
}, function() {
});

client.deviceStates.update({deviceName: 'Stasis:42',
                                            deviceState: 'NOT_INUSE'}, function() {});

The important thing is to align subsequent lines in a way that flows better and is easier to read.

Note: github didn't indent my comment as it should have. :(

samuelg commented on tests/sla_tests.js:
```diff
@@ -429,4 +465,180 @@ describe('SLA Bridge and Channels Tester', function() {
+    function markAsRinging() {
+      setTimeout(function() {
+        if (isMixing && bridges.length === 1 && inbound.inbound &&
+          dialed[0] && dialed[0].wasAnswered && inbound.wasAnswered &&
```
dialed[0] should be aligned with isMixing and so forth so this statement is easier to read.

samuelg commented:
+1 on everything Matt said. Vertical spacing would help a lot in making your code more readable.

TACambron committed:
Added spacing for readability and a separate update function.

Fixed many alignment issues with indentation and vertical spacing
to make the code more readable. Created a function to perform the
state update instead of creating an update variable each time an
update was required. Modified the errHandler function for the tests
by moving the specific error handlings to the tests that required
them. Removed duplicate tests. Changed the noEndpoints variable
to noStations, which is what was originally intended to be there.

TACambron committed:
Created references for device states.

Created constant references for device states and added lodash to
package.json.

samuelg commented:
Looks good. Make sure you squash the commits before merging.

TACambron commented:
Squashed into #18
